boards/xtensa/esp32/esp32-devkitc/src/esp32_ws2812.c

```diff
@@ -0,0 +1,333 @@
+  {
+    lederr("DATA is NULL\n");
+    return 0;
+  }
+  
+  nxsem_wait(&dev_data->exclsem);
+
+  uint32_t mem = priv->rmt_dev->start_address;
+  uint32_t used_words = 0;
+  uint32_t current_pixel = 0;
+  uint8_t *pixel_ptr = (uint8_t *)data;
+  uint8_t pixel;
+  uint32_t reg1_addr = RMT_CHnCONF1_REG(priv->rmt_dev->channel);
+  modifyreg32(reg1_addr, 0, RMT_MEM_RD_RST_CHn);
+  modifyreg32(reg1_addr, RMT_MEM_RD_RST_CHn, 0);
+
+  while(used_words<priv->rmt_dev->available_words)
+    {
+      if (current_pixel < len)
+        {
+          pixel = *pixel_ptr++;
+          for (int i = 0; i < 8; i++)
+            {
+              int bit = pixel & 0x80L;
+              if (bit)
+                {
+                  putreg32(LOGIC_ONE, mem);
+                }
+              else
+               {
+                 putreg32(LOGIC_ZERO, mem);
+               }
+              mem += 4;
+              used_words++;
+              pixel <<= 1;
+            }
+          current_pixel++;
+        } 
+      else 
+        {
+          putreg32(0, mem); 
+          break;
+        }
+    }
+  
+  if(used_words<priv->rmt_dev->available_words) 
+    {
+      /* Small transactions won't need a the swap buffer IRQ */
+      modifyreg32(
+        RMT_INT_ENA_REG, 
+        RMT_CHn_TX_THR_EVENT_INT_ENA(priv->rmt_dev->channel), 
+        RMT_CHn_TX_END_INT_ENA(priv->rmt_dev->channel));
+    } 
+  else 
+    {
+    /* In large transactions inform the current state to the IRQ */
+    priv->rmt_dev->next_pixel_to_load = current_pixel;
+    priv->rmt_dev->data_ptr = (uint8_t*) data;
+    priv->rmt_dev->data_len = len;
+    priv->rmt_dev->frame = 0;
+    
+    /* Enable memory wrap around */
+    modifyreg32(RMT_APB_CONF_REG, 0, BIT(1));    
+
+    /* Enable buffering IRQ */
+    modifyreg32(
+      RMT_INT_ENA_REG, 
+      0, 
+      RMT_CHn_TX_END_INT_ENA(priv->rmt_dev->channel) | 
+      RMT_CHn_TX_THR_EVENT_INT_ENA(priv->rmt_dev->channel));
+  }
+
+  /* Start the transaction */
+  modifyreg32(reg1_addr, 0, RMT_TX_START_CHn);
+  nxsem_wait(&priv->rmt_dev->tx_sem);
+
+  /* Wait for the TX lock and release it after */
+  nxsem_wait(&priv->rmt_dev->tx_sem); 
+  nxsem_post(&priv->rmt_dev->tx_sem);
+  nxsem_post(&dev_data->exclsem);
+
+  return len;
+}
+/****************************************************************************
+ * Name: board_ws2812_initialize
+ *
+ * Description:
+ *   Initialize and register the WS2812 LED driver using the RMT 
+ *
+ * Input Parameters:
+ *   devno - The device number, used to build the device path as /dev/ledsN
+ *   rmt_dev - Pointer to the RMT device that will be used
+ *   nleds - number of LEDs
+ *
+ * Returned Value:
+ *   Zero (OK) on success; a negated errno value on failure.
+ *
+ ****************************************************************************/
+
+int board_ws2812_initialize(int devno, uint16_t nleds, struct rmt_dev_s *rmt_dev)
+{
+  char devpath[13];
+  int ret;
+
+  FAR struct ws2812_dev_s *dev_data;
```

Comment:
```suggestion
  struct ws2812_dev_s *dev_data;
```
`FAR` qualifier should not be used in arch-specific code.

Reply:
Are these supposed to be removed as well?

```
static int ws2812_rmt_open(FAR struct file *filep)
{
  FAR struct inode *inode = filep->f_inode;
  FAR struct ws2812_dev_s *dev_data = inode->i_private;
  FAR struct ws2812_rmt_s *priv = (FAR struct ws2812_rmt_s *)
                                      dev_data->private;

```

Reply:
Yes. `FAR` and `CODE` have been wiped from every arch implementation some time ago, since they have an empty definition for newer archs and just contribute to code bloat.